In [1]:
import pandas as pd
import numpy as np

### Generating the Data

In [2]:
data = pd.read_csv('https://raw.githubusercontent.com/edyoda/Data-Scientist-program/master/Practice%20Problems/data/yellow-small%2Badult-stretch.data',names=['color','size','act','age','inflated'])

In [3]:
data

color   size      act    age inflated
0   YELLOW  SMALL  STRETCH  ADULT        T
1   YELLOW  SMALL  STRETCH  CHILD        T
2   YELLOW  SMALL      DIP  ADULT        T
3   YELLOW  SMALL      DIP  CHILD        T
4   YELLOW  LARGE  STRETCH  ADULT        T
5   YELLOW  LARGE  STRETCH  CHILD        F
6   YELLOW  LARGE      DIP  ADULT        F
7   YELLOW  LARGE      DIP  CHILD        F
8   PURPLE  SMALL  STRETCH  ADULT        T
9   PURPLE  SMALL  STRETCH  CHILD        F
10  PURPLE  SMALL      DIP  ADULT        F
11  PURPLE  SMALL      DIP  CHILD        F
12  PURPLE  LARGE  STRETCH  ADULT        T
13  PURPLE  LARGE  STRETCH  CHILD        F
14  PURPLE  LARGE      DIP  ADULT        F
15  PURPLE  LARGE      DIP  CHILD        F

In [4]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 16 entries, 0 to 15
Data columns (total 5 columns):
color       16 non-null object
size        16 non-null object
act         16 non-null object
age         16 non-null object
inflated    16 non-null object
dtypes: object(5)
memory usage: 768.0+ bytes


In [5]:
### Finding the Entropy of Each  column

In [6]:
data.inflated.value_counts()

F    9
T    7
Name: inflated, dtype: int64

### Total  Entropy of the inflated column

In [7]:
Entropy_inflated  = -(7/16)*np.log2(7/16)- (9/16)*np.log2(9/16)
Entropy_inflated

0.9886994082884974

In [8]:
data.columns[0]

'color'

In [9]:
data.color.value_counts()

PURPLE    8
YELLOW    8
Name: color, dtype: int64

### Entropy of the column color

In [10]:
data.groupby("color").inflated.value_counts()

color   inflated
PURPLE  F           6
        T           2
YELLOW  T           5
        F           3
Name: inflated, dtype: int64

In [11]:
Entropy_color_purple = -(6/8)*np.log2(6/8)-(2/8)*np.log2(2/8)
Entropy_color_purple

0.8112781244591328

In [12]:
data.groupby("color").inflated.value_counts()

color   inflated
PURPLE  F           6
        T           2
YELLOW  T           5
        F           3
Name: inflated, dtype: int64

In [13]:
Entropy_color_yellow = -(5/8)*np.log2(5/8)-(3/8)*np.log2(3/8)
Entropy_color_yellow

0.954434002924965

### Information Gain of the column "color", Inflated

In [14]:
Ingormation_gain_color = Entropy_inflated-(8/16)*Entropy_color_yellow-(8/16)*Entropy_color_purple
abs(Ingormation_gain_color)

0.10584334459644845

In [15]:
### Entropy of the  column "Size"

In [16]:
data.groupby("size").inflated.value_counts()

size   inflated
LARGE  F           6
       T           2
SMALL  T           5
       F           3
Name: inflated, dtype: int64

In [17]:
Entropy_size_large = -(2/8)*np.log2(2/8)-(6/8)*np.log2(6/8)
Entropy_size_large

0.8112781244591328

In [18]:
Entropy_size_small = -(5/8)*np.log2(5/8)-(3/8)*np.log2(3/8)
Entropy_size_small

0.954434002924965

### Information gain for the column "Size"

In [19]:
Ingormation_gain_size = Entropy_inflated-(8/16)*Entropy_size_large-(8/16)*Entropy_size_small
abs(Ingormation_gain_size)

0.10584334459644845

### Entropy of the column "act"

In [20]:
data.groupby("act").inflated.value_counts()

act      inflated
DIP      F           6
         T           2
STRETCH  T           5
         F           3
Name: inflated, dtype: int64

In [21]:
Entropy_size_dip = -(2/8)*np.log2(2/8)-(6/8)*np.log2(6/8)
Entropy_size_dip

0.8112781244591328

In [22]:
Entropy_size_stretch = -(5/8)*np.log2(5/8)-(3/8)*np.log2(3/8)
Entropy_size_stretch


0.954434002924965

### Information Gain For column "Act"

In [23]:
Ingormation_gain_act = Entropy_inflated-(8/16)*Entropy_size_dip-(8/16)*Entropy_size_stretch
abs(Ingormation_gain_act)

0.10584334459644845

### Entropy of the column "Age"

In [24]:
data.groupby("age").inflated.value_counts()

age    inflated
ADULT  T           5
       F           3
CHILD  F           6
       T           2
Name: inflated, dtype: int64

In [25]:
Entropy_size_adult = -(5/8)*np.log2(5/8)-(3/8)*np.log2(3/8)
Entropy_size_adult


0.954434002924965

In [26]:
Entropy_size_child = -(2/8)*np.log2(2/8)-(6/8)*np.log2(6/8)
Entropy_size_child

0.8112781244591328

### Information Gain for the column "Age"

In [27]:
Ingormation_gain_age = Entropy_inflated-(8/16)*Entropy_size_adult-(8/16)*Entropy_size_child
abs(Ingormation_gain_age)

0.10584334459644845

### Information Gain for columns:
1. color : 0.10584334459644845
2. size : 0.10584334459644845
3. Act : 0.10584334459644845
4. Age : 0.10584334459644845
    

<img src = "entropy.jpg"/>

### Conclusion of the root node i.e Color

### Spliting the next node

In [28]:
data[data.color== "YELLOW"]

color   size      act    age inflated
0  YELLOW  SMALL  STRETCH  ADULT        T
1  YELLOW  SMALL  STRETCH  CHILD        T
2  YELLOW  SMALL      DIP  ADULT        T
3  YELLOW  SMALL      DIP  CHILD        T
4  YELLOW  LARGE  STRETCH  ADULT        T
5  YELLOW  LARGE  STRETCH  CHILD        F
6  YELLOW  LARGE      DIP  ADULT        F
7  YELLOW  LARGE      DIP  CHILD        F

#### Entropy of the Color column category i.e  is Yellow

In [29]:
Entropy_color_yellow


0.954434002924965

### Information Gain for the next node for decision tree i.e with size column

In [30]:
Entropy_color_yellow- (0.5)* (-(1)*np.log2(1))-(0.5)*(-(1/4)*np.log2(1/4)-(3/4)*np.log2(3/4))

0.5487949406953987

### Information Gain for the next node for decision tree i.e with act column

In [31]:
Entropy_color_yellow- ((0.5)* (-(0.5)*np.log2(0.5)-(0.5)*np.log2(0.5)))-((0.5)*(-(1/4)*np.log2(1/4)-(3/4)*np.log2(3/4)))

0.048794940695398636

### Information Gain for the next node for decision tree i.e with Age column

In [32]:
Entropy_color_yellow- ((0.5)* (-(0.5)*np.log2(0.5)-(0.5)*np.log2(0.5)))-((0.5)*(-(1/4)*np.log2(1/4)-(3/4)*np.log2(3/4)))

0.048794940695398636

Information Gain for the Color column:
    1. size: 0.54
    2. act : 0.048
    3. Age : 0.048

### Conclusion
1. Size column is selected as the next split.  Because of the inforamtion gain

<img src = "data.png"/>

### Information Gain on the color attribute category i.e is "Purple"

In [33]:
data[data.color== "PURPLE"]

color   size      act    age inflated
8   PURPLE  SMALL  STRETCH  ADULT        T
9   PURPLE  SMALL  STRETCH  CHILD        F
10  PURPLE  SMALL      DIP  ADULT        F
11  PURPLE  SMALL      DIP  CHILD        F
12  PURPLE  LARGE  STRETCH  ADULT        T
13  PURPLE  LARGE  STRETCH  CHILD        F
14  PURPLE  LARGE      DIP  ADULT        F
15  PURPLE  LARGE      DIP  CHILD        F

In [34]:
Entropy_color_purple

0.8112781244591328

### Information Gain for the next node for decision tree i.e with size column

In [35]:
Entropy_color_purple-((0.5)* (-(1/4)*np.log2(1/4)-(3/4)*np.log2(3/4)))- ((0.5)* (-(1/4)*np.log2(1/4)-(3/4)*np.log2(3/4)))

0.0

### Information Gain for the next node for decision tree i.e with act column

In [36]:
Entropy_color_purple-((0.5)* (-(2/4)*np.log2(2/4)-(2/4)*np.log2(2/4)))- ((0.5)* (-(1)*np.log2(1)))

0.31127812445913283

### Information Gain for the next node for decision tree i.e with Age column

In [37]:
Entropy_color_purple-((0.5)* (-(2/4)*np.log2(2/4)-(2/4)*np.log2(2/4)))- ((0.5)* (-(1)*np.log2(1)))

0.31127812445913283

In [38]:
data[data.color== "PURPLE"]

color   size      act    age inflated
8   PURPLE  SMALL  STRETCH  ADULT        T
9   PURPLE  SMALL  STRETCH  CHILD        F
10  PURPLE  SMALL      DIP  ADULT        F
11  PURPLE  SMALL      DIP  CHILD        F
12  PURPLE  LARGE  STRETCH  ADULT        T
13  PURPLE  LARGE  STRETCH  CHILD        F
14  PURPLE  LARGE      DIP  ADULT        F
15  PURPLE  LARGE      DIP  CHILD        F

Conclusion :
    Since the information gain for Age and  Act attribute is same  and it is highest we select the Act as the next node or the leaf node

<img src = "decision.png"/>

<img src ="names.drawio"/>

In [39]:
from sklearn.tree import ExtraTreeClassifier, export_graphviz, DecisionTreeClassifier

In [40]:
dt =  DecisionTreeClassifier(criterion="entropy")

In [41]:
from sklearn.model_selection import train_test_split

In [48]:
trainX, testX, trainY, testY = train_test_split(data.drop(columns = "inflated"), data.inflated)

In [43]:
data.columns

Index(['color', 'size', 'act', 'age', 'inflated'], dtype='object')

In [47]:
from sklearn.preprocessing import LabelEncoder
for col in data.columns:
    le = LabelEncoder()
    data[col] = le.fit_transform(data[col])

In [45]:
data

color  size  act  age  inflated
0       1     1    1    0         1
1       1     1    1    1         1
2       1     1    0    0         1
3       1     1    0    1         1
4       1     0    1    0         1
5       1     0    1    1         0
6       1     0    0    0         0
7       1     0    0    1         0
8       0     1    1    0         1
9       0     1    1    1         0
10      0     1    0    0         0
11      0     1    0    1         0
12      0     0    1    0         1
13      0     0    1    1         0
14      0     0    0    0         0
15      0     0    0    1         0

In [49]:
clf = dt.fit(trainX,trainY)

In [50]:
export_graphviz(dt, 'dt.tree', feature_names=['color','act','size','Age'])

In [51]:
data.drop(columns = "inflated")

color  size  act  age
0       1     1    1    0
1       1     1    1    1
2       1     1    0    0
3       1     1    0    1
4       1     0    1    0
5       1     0    1    1
6       1     0    0    0
7       1     0    0    1
8       0     1    1    0
9       0     1    1    1
10      0     1    0    0
11      0     1    0    1
12      0     0    1    0
13      0     0    1    1
14      0     0    0    0
15      0     0    0    1

In [52]:
help(export_graphviz)

Help on function export_graphviz in module sklearn.tree.export:

export_graphviz(decision_tree, out_file=None, max_depth=None, feature_names=None, class_names=None, label='all', filled=False, leaves_parallel=False, impurity=True, node_ids=False, proportion=False, rotate=False, rounded=False, special_characters=False, precision=3)
    Export a decision tree in DOT format.
    
    This function generates a GraphViz representation of the decision tree,
    which is then written into `out_file`. Once exported, graphical renderings
    can be generated using, for example::
    
        $ dot -Tps tree.dot -o tree.ps      (PostScript format)
        $ dot -Tpng tree.dot -o tree.png    (PNG format)
    
    The sample counts that are shown are weighted with any sample_weights that
    might be present.
    
    Read more in the :ref:`User Guide <tree>`.
    
    Parameters
    ----------
    decision_tree : decision tree classifier
        The decision tree to be exported to GraphViz.
    
 

### Gini Gain

In [53]:
gini_data = pd.read_csv('https://raw.githubusercontent.com/edyoda/Data-Scientist-program/master/Practice%20Problems/data/yellow-small%2Badult-stretch.data',names=['color','size','act','age','inflated'])

In [54]:
gini_data

color   size      act    age inflated
0   YELLOW  SMALL  STRETCH  ADULT        T
1   YELLOW  SMALL  STRETCH  CHILD        T
2   YELLOW  SMALL      DIP  ADULT        T
3   YELLOW  SMALL      DIP  CHILD        T
4   YELLOW  LARGE  STRETCH  ADULT        T
5   YELLOW  LARGE  STRETCH  CHILD        F
6   YELLOW  LARGE      DIP  ADULT        F
7   YELLOW  LARGE      DIP  CHILD        F
8   PURPLE  SMALL  STRETCH  ADULT        T
9   PURPLE  SMALL  STRETCH  CHILD        F
10  PURPLE  SMALL      DIP  ADULT        F
11  PURPLE  SMALL      DIP  CHILD        F
12  PURPLE  LARGE  STRETCH  ADULT        T
13  PURPLE  LARGE  STRETCH  CHILD        F
14  PURPLE  LARGE      DIP  ADULT        F
15  PURPLE  LARGE      DIP  CHILD        F

In [55]:
from sklearn.preprocessing import OrdinalEncoder
oh = OrdinalEncoder()
feature_target= oh.fit_transform(gini_data)

In [56]:
feature_target

array([[1., 1., 1., 0., 1.],
       [1., 1., 1., 1., 1.],
       [1., 1., 0., 0., 1.],
       [1., 1., 0., 1., 1.],
       [1., 0., 1., 0., 1.],
       [1., 0., 1., 1., 0.],
       [1., 0., 0., 0., 0.],
       [1., 0., 0., 1., 0.],
       [0., 1., 1., 0., 1.],
       [0., 1., 1., 1., 0.],
       [0., 1., 0., 0., 0.],
       [0., 1., 0., 1., 0.],
       [0., 0., 1., 0., 1.],
       [0., 0., 1., 1., 0.],
       [0., 0., 0., 0., 0.],
       [0., 0., 0., 1., 0.]])

In [57]:
data.columns

Index(['color', 'size', 'act', 'age', 'inflated'], dtype='object')

In [58]:
new_data = pd.DataFrame(feature_target,columns= ['color', 'size', 'act', 'age', 'inflated'])

In [59]:
new_data


color  size  act  age  inflated
0     1.0   1.0  1.0  0.0       1.0
1     1.0   1.0  1.0  1.0       1.0
2     1.0   1.0  0.0  0.0       1.0
3     1.0   1.0  0.0  1.0       1.0
4     1.0   0.0  1.0  0.0       1.0
5     1.0   0.0  1.0  1.0       0.0
6     1.0   0.0  0.0  0.0       0.0
7     1.0   0.0  0.0  1.0       0.0
8     0.0   1.0  1.0  0.0       1.0
9     0.0   1.0  1.0  1.0       0.0
10    0.0   1.0  0.0  0.0       0.0
11    0.0   1.0  0.0  1.0       0.0
12    0.0   0.0  1.0  0.0       1.0
13    0.0   0.0  1.0  1.0       0.0
14    0.0   0.0  0.0  0.0       0.0
15    0.0   0.0  0.0  1.0       0.0

In [60]:
df = DecisionTreeClassifier(criterion= 'gini')

In [61]:
df.fit(new_data.drop(columns= "inflated"), new_data.inflated)

DecisionTreeClassifier(class_weight=None, criterion='gini', max_depth=None,
                       max_features=None, max_leaf_nodes=None,
                       min_impurity_decrease=0.0, min_impurity_split=None,
                       min_samples_leaf=1, min_samples_split=2,
                       min_weight_fraction_leaf=0.0, presort=False,
                       random_state=None, splitter='best')

In [62]:
new_data.columns

Index(['color', 'size', 'act', 'age', 'inflated'], dtype='object')

In [63]:
export_graphviz(df, 'df-gini.tree',feature_names = ['color', 'size', 'act', 'age' ])

<img src = 'ginigain.jpg'/>

In [64]:
gini_data

color   size      act    age inflated
0   YELLOW  SMALL  STRETCH  ADULT        T
1   YELLOW  SMALL  STRETCH  CHILD        T
2   YELLOW  SMALL      DIP  ADULT        T
3   YELLOW  SMALL      DIP  CHILD        T
4   YELLOW  LARGE  STRETCH  ADULT        T
5   YELLOW  LARGE  STRETCH  CHILD        F
6   YELLOW  LARGE      DIP  ADULT        F
7   YELLOW  LARGE      DIP  CHILD        F
8   PURPLE  SMALL  STRETCH  ADULT        T
9   PURPLE  SMALL  STRETCH  CHILD        F
10  PURPLE  SMALL      DIP  ADULT        F
11  PURPLE  SMALL      DIP  CHILD        F
12  PURPLE  LARGE  STRETCH  ADULT        T
13  PURPLE  LARGE  STRETCH  CHILD        F
14  PURPLE  LARGE      DIP  ADULT        F
15  PURPLE  LARGE      DIP  CHILD        F

In [81]:
list(new_data.color.value_counts().index)

[0.0, 1.0]

In [82]:
new_data.inflated.value_counts()

0.0    9
1.0    7
Name: inflated, dtype: int64

In [84]:
Ginibeforesplit = ((7/16)*(1-(7/16)))+((9/16)*(1-(9/16)))
Ginibeforesplit
    
        

0.4921875


How is the data discretized by Decision Tree?
This is done using Binning Techniques

For color column value ranges from [0,1]

Buckets [.,0.5.]
For act column

Buckets [.,0.5,.]
For age columns

Buckets [.,0.5,.]
For size column

Buckets [.,0.5,.]
What are the possible questions or decisions?
For color attribute, questions can be any of the following
(a) color < 0.5
For act attribute, questions can be
(c) act < 0.5, 
For size
(e) size < 0.5
For age
(f) age < 0.5

In [ ]:
Try for the color column which is less than 0.5


In [87]:
new_data[new_data.color<0.5]

color  size  act  age  inflated
8     0.0   1.0  1.0  0.0       1.0
9     0.0   1.0  1.0  1.0       0.0
10    0.0   1.0  0.0  0.0       0.0
11    0.0   1.0  0.0  1.0       0.0
12    0.0   0.0  1.0  0.0       1.0
13    0.0   0.0  1.0  1.0       0.0
14    0.0   0.0  0.0  0.0       0.0
15    0.0   0.0  0.0  1.0       0.0

In [89]:
Ginileft = ((2/8)*(1-(2/8)))+((6/8)*(1-(6/8)))
Ginileft

0.375

In [90]:
new_data[new_data.color>0.5]

color  size  act  age  inflated
0    1.0   1.0  1.0  0.0       1.0
1    1.0   1.0  1.0  1.0       1.0
2    1.0   1.0  0.0  0.0       1.0
3    1.0   1.0  0.0  1.0       1.0
4    1.0   0.0  1.0  0.0       1.0
5    1.0   0.0  1.0  1.0       0.0
6    1.0   0.0  0.0  0.0       0.0
7    1.0   0.0  0.0  1.0       0.0

In [91]:
Giniright = ((5/8)*(1-(5/8)))+((3/8)*(1-(3/8)))

In [92]:
Giniright

0.46875

In [93]:
Ginisplit = (8/16)*Ginileft +(8/16)*Giniright

In [94]:
Ginigain_color = Ginibeforesplit- Ginisplit

In [95]:
Ginigain_color

0.0703125

### GiniGain for the column name Act

In [96]:
new_data[new_data.act<0.5]

color  size  act  age  inflated
2     1.0   1.0  0.0  0.0       1.0
3     1.0   1.0  0.0  1.0       1.0
6     1.0   0.0  0.0  0.0       0.0
7     1.0   0.0  0.0  1.0       0.0
10    0.0   1.0  0.0  0.0       0.0
11    0.0   1.0  0.0  1.0       0.0
14    0.0   0.0  0.0  0.0       0.0
15    0.0   0.0  0.0  1.0       0.0

In [97]:
Ginileft_act = ((2/8)*(1-(2/8)))+((6/8)*(1-(6/8)))

In [98]:
GiniRight_act =((5/8)*(1-(5/8)))+((3/8)*(1-(3/8))) 

In [99]:
Ginisplit_act =  (8/16)*Ginileft +(8/16)*Giniright

In [104]:
Ginigain_act = Ginibeforesplit - Ginisplit_act
Ginigain_act

0.0703125

### conclusion Size is selected as the root node because of high gini split

Calculating for the sub tree

In [111]:
gini_data.sample()

color   size      act    age inflated
1  YELLOW  SMALL  STRETCH  CHILD        T

In [118]:

gini_data[gini_data['size']== 'LARGE']

color   size      act    age inflated
4   YELLOW  LARGE  STRETCH  ADULT        T
5   YELLOW  LARGE  STRETCH  CHILD        F
6   YELLOW  LARGE      DIP  ADULT        F
7   YELLOW  LARGE      DIP  CHILD        F
12  PURPLE  LARGE  STRETCH  ADULT        T
13  PURPLE  LARGE  STRETCH  CHILD        F
14  PURPLE  LARGE      DIP  ADULT        F
15  PURPLE  LARGE      DIP  CHILD        F

In [119]:
ginibeforesplit1 = ((2/8)*(1-(2/8)))+((5/8)*(1-(5/8))) 

#### Gini Gain for all the columns

Gini gain for the column "color"

In [122]:
ginileftcolor1=((1/4)*(1-(1/4)))+((3/4)*(1-(3/4))) 

In [124]:
ginirightcolor1 = ((1/4)*(1-(1/4)))+((3/4)*(1-(3/4))) 

In [127]:
ginisplit1_color = (.5*ginileftcolor1) + (0.5*ginirightcolor1)
ginisplit1_color

0.375

In [136]:
subdata = gini_data[gini_data['size']== 'LARGE']

In [137]:
subdata

color   size      act    age inflated
4   YELLOW  LARGE  STRETCH  ADULT        T
5   YELLOW  LARGE  STRETCH  CHILD        F
6   YELLOW  LARGE      DIP  ADULT        F
7   YELLOW  LARGE      DIP  CHILD        F
12  PURPLE  LARGE  STRETCH  ADULT        T
13  PURPLE  LARGE  STRETCH  CHILD        F
14  PURPLE  LARGE      DIP  ADULT        F
15  PURPLE  LARGE      DIP  CHILD        F

In [ ]:
Gini gain for the column 'Age'

In [130]:
ginileftage1 = ((2/4)*(1-(2/4)))+((2/4)*(1-(2/4))) 


In [138]:
ginirightage1 =((4/4)*(1-(4/4))) 


In [139]:
ginisplit1 = 0.5*(ginileftage1) + 0.5*(ginirightage1)

ginisplit1

0.25

In [143]:
subdata1 = gini_data[gini_data['size']== 'SMALL']
subdata1

color   size      act    age inflated
0   YELLOW  SMALL  STRETCH  ADULT        T
1   YELLOW  SMALL  STRETCH  CHILD        T
2   YELLOW  SMALL      DIP  ADULT        T
3   YELLOW  SMALL      DIP  CHILD        T
8   PURPLE  SMALL  STRETCH  ADULT        T
9   PURPLE  SMALL  STRETCH  CHILD        F
10  PURPLE  SMALL      DIP  ADULT        F
11  PURPLE  SMALL      DIP  CHILD        F

In [144]:
ginileft2 = ((1/4)*(1-(1/4)))+((3/4)*(1-(3/4))) 


In [145]:
giniright2 =  ((2/4)*(1-(2/4)))+((2/4)*(1-(2/4))) 

In [146]:
ginisplit2 = 0.5*(ginileft2) + 0.5*(giniright2)

ginisplit2

0.4375